File: plot_benchmark_comparison.R
#!/usr/bin/env Rscript
# ---------------------------------------------------------------------------
# Per-baseline comparison figures for the four-way benchmark, in R/ggplot2.
#
# The Python side (compare_transfer_enformer.py) produces this figure family for
# Enformer only. This script produces the same family for ALL THREE baselines —
# pretrained Borzoi, Enformer and AlphaGenome — from the single reduced table
# results_benchmark/benchmark_comparison_{split}.csv, so every panel is drawn
# from the same numbers on the same shared 896 x 128 bp grid.
#
# Six figures per baseline:
#   violin        distribution of per-track R, transfer vs baseline
#   ecdf          cumulative distribution of both
#   scatter       per-target transfer vs baseline, with y = x
#   delta_hist    histogram of the paired difference
#   paired        per-target dumbbell, sorted by baseline R
#   plausibility  is the matched track the same TF, a relative, or unrelated?
#
# On ACCRE:
#   module load r
#   export R_LIBS_USER=~/R/x86_64-pc-linux-gnu-library/4.6
#   Rscript plot_benchmark_comparison.R
# In RStudio Server: open the project and source() this file.
# ---------------------------------------------------------------------------

suppressPackageStartupMessages({
  need <- c("ggplot2", "dplyr", "tidyr", "scales")
  miss <- need[!vapply(need, requireNamespace, logical(1), quietly = TRUE)]
  if (length(miss)) {
    stop("Missing R packages: ", paste(miss, collapse = ", "), "\n",
         "  install.packages(c(", paste0('"', miss, '"', collapse = ", "), "))",
         call. = FALSE)
  }
  library(ggplot2); library(dplyr); library(tidyr); library(scales)
})

# --- config ----------------------------------------------------------------
args     <- commandArgs(trailingOnly = TRUE)
split    <- if (length(args) >= 1) args[1] else "test"
in_csv   <- file.path("results_benchmark", sprintf("benchmark_comparison_%s.csv", split))
out_dir  <- "results_comparison_R"
dir.create(out_dir, showWarnings = FALSE)

# validated categorical palette (same hues as the Python figures, so the two
# sets can sit side by side in one deck without reading as different studies)
C_TL <- "#2a78d6"; C_BZ <- "#eb6834"; C_AG <- "#1baf7a"; C_EF <- "#4a3aa7"
INK  <- "#0b0b0b"; INK2 <- "#52514e"; GRID <- "#d9d8d4"; SURF <- "#fcfcfb"

BASELINES <- list(
  list(key = "borzoi_pretrained", label = "Borzoi pretrained", colour = C_BZ),
  list(key = "enformer",          label = "Enformer",          colour = C_EF),
  list(key = "alphagenome",       label = "AlphaGenome",       colour = C_AG)
)

theme_bench <- function() {
  theme_minimal(base_size = 11) +
    theme(
      plot.background   = element_rect(fill = SURF, colour = NA),
      panel.background  = element_rect(fill = SURF, colour = NA),
      panel.grid.minor  = element_blank(),
      panel.grid.major  = element_line(colour = GRID, linewidth = 0.3),
      axis.text         = element_text(colour = INK2),
      axis.title        = element_text(colour = INK),
      plot.title        = element_text(colour = INK, face = "bold", size = 12.5),
      plot.subtitle     = element_text(colour = INK2, size = 9.5),
      legend.position   = "top",
      legend.title      = element_blank(),
      legend.text       = element_text(colour = INK)
    )
}

save_both <- function(p, name, w = 7.2, h = 4.8) {
  for (ext in c("png", "pdf")) {
    ggsave(file.path(out_dir, paste0(name, "_", split, ".", ext)),
           p, width = w, height = h, dpi = 200, bg = SURF)
  }
}

# TF name out of a matched-track description. Borzoi and Enformer use
# "CHIP:<factor>:<cell/tissue>"; AlphaGenome uses "<factor> (<biosample>)".
tf_of <- function(x, key) {
  if (key == "alphagenome") toupper(trimws(sub("\\s*\\(.*$", "", x)))
  else                      toupper(trimws(vapply(strsplit(x, ":"), function(p)
                              if (length(p) >= 2) p[2] else "", character(1))))
}

# Crude TF-family grouping, mirroring borzoi_baseline_utils._family so the R and
# Python plausibility panels agree.
FAMS <- c("FOX","GATA","HOX","SOX","KLF","STAT","IRF","SMAD","TCF","LEF","NFK",
          "NR","ZNF","E2F","ETS","ELF","CEBP","JUN","FOS","MYC","RUNX","TEAD","TP","SP")
family_of <- function(f) {
  f <- toupper(f)
  hit <- FAMS[vapply(FAMS, function(p) startsWith(f, p), logical(1))]
  if (length(hit)) hit[which.max(nchar(hit))] else substr(f, 1, 3)
}

classify <- function(target_tf, match_tf) {
  if (is.na(match_tf) || match_tf == "" || match_tf == "NAN") return("unclear")
  if (target_tf != "" && target_tf == match_tf)               return("same TF")
  if (target_tf != "" && family_of(target_tf) == family_of(match_tf)) return("related family")
  "different TF"
}

# --- data ------------------------------------------------------------------
if (!file.exists(in_csv)) stop("missing ", in_csv, " — run benchmark_compare_all.py first")
d <- read.csv(in_csv, stringsAsFactors = FALSE)
cat(sprintf("Loaded %d targets from %s\n", nrow(d), in_csv))

summary_rows <- list()

for (b in BASELINES) {
  key <- b$key; lab <- b$label; col <- b$colour
  rcol <- paste0(key, "_r"); tcol <- paste0(key, "_track")

  df <- data.frame(
    target      = d$target_factor,
    cell        = d$target_cell_line,
    transfer_r  = d$borzoi_tl_r,
    baseline_r  = d[[rcol]],
    match_track = d[[tcol]],
    stringsAsFactors = FALSE
  )
  df$delta  <- df$transfer_r - df$baseline_r
  df$winner <- ifelse(abs(df$delta) <= 0.01, "tie",
                      ifelse(df$delta > 0, "Borzoi-TL", lab))

  pal <- c("Borzoi-TL" = C_TL); pal[lab] <- col

  # 1. violin + box ---------------------------------------------------------
  long <- df |>
    select(transfer_r, baseline_r) |>
    pivot_longer(everything(), names_to = "model", values_to = "r") |>
    mutate(model = factor(ifelse(model == "transfer_r", "Borzoi-TL", lab),
                          levels = c("Borzoi-TL", lab)))
  means <- long |> group_by(model) |> summarise(m = mean(r), .groups = "drop")

  p1 <- ggplot(long, aes(model, r, fill = model, colour = model)) +
    geom_violin(alpha = 0.18, linewidth = 0.7, width = 0.85) +
    geom_boxplot(width = 0.16, outlier.shape = NA, fill = "white", linewidth = 0.7) +
    # pinned above the violin, not at the mean, so it never covers the box
    geom_text(data = means, aes(model, 0.985, label = sprintf("%.3f", m)),
              vjust = 1, fontface = "bold", size = 4.0, show.legend = FALSE) +
    scale_fill_manual(values = pal) + scale_colour_manual(values = pal) +
    labs(title = sprintf("Per-track accuracy: Borzoi-TL vs %s", lab),
         subtitle = sprintf("%d held-out CRC TF ChIP-seq tracks · shared 896 x 128 bp grid · bold = mean", nrow(df)),
         x = NULL, y = "Pearson R per target track") +
    ylim(0, 1.02) + theme_bench() + theme(legend.position = "none")
  save_both(p1, paste0(key, "_violin"))

  # 2. ECDF -----------------------------------------------------------------
  p2 <- ggplot(long, aes(r, colour = model)) +
    stat_ecdf(linewidth = 1.1) +
    scale_colour_manual(values = pal) +
    labs(title = sprintf("Cumulative distribution of per-track R"),
         subtitle = sprintf("a curve further right is better · Borzoi-TL vs %s", lab),
         x = "Pearson R", y = "fraction of tracks at or below") +
    theme_bench()
  save_both(p2, paste0(key, "_ecdf"))

  # 3. scatter --------------------------------------------------------------
  p3 <- ggplot(df, aes(baseline_r, transfer_r)) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", colour = INK2, linewidth = 0.5) +
    geom_point(aes(colour = winner), size = 2.1, alpha = 0.8) +
    scale_colour_manual(values = c(pal, tie = INK2)) +
    coord_equal(xlim = c(0.1, 0.95), ylim = c(0.1, 0.95)) +
    labs(title = sprintf("Per-target comparison vs %s", lab),
         subtitle = "above the dashed line = Borzoi-TL better",
         x = sprintf("%s best-match R", lab), y = "Borzoi-TL R") +
    theme_bench()
  save_both(p3, paste0(key, "_scatter"), w = 6.0, h = 6.0)

  # 4. paired-difference histogram -----------------------------------------
  mu <- mean(df$delta)
  p4 <- ggplot(df, aes(delta)) +
    geom_histogram(bins = 34, fill = col, colour = SURF, alpha = 0.85, linewidth = 0.3) +
    geom_vline(xintercept = 0, colour = INK2, linewidth = 0.6) +
    geom_vline(xintercept = mu, colour = C_TL, linewidth = 1.0, linetype = "dashed") +
    annotate("text", x = mu, y = Inf, vjust = 1.8, hjust = -0.08,
             label = sprintf("mean %+.3f", mu), colour = C_TL, fontface = "bold", size = 3.6) +
    labs(title = sprintf("Paired difference vs %s", lab),
         subtitle = sprintf("%d Borzoi-TL wins · %d losses · %d ties (|ΔR| ≤ 0.01)",
                            sum(df$winner == "Borzoi-TL"), sum(df$winner == lab), sum(df$winner == "tie")),
         x = "ΔR  (Borzoi-TL minus baseline)", y = "number of target tracks") +
    theme_bench()
  save_both(p4, paste0(key, "_delta_hist"))

  # 5. paired dumbbell ------------------------------------------------------
  ord <- df |> arrange(baseline_r) |> mutate(i = row_number())
  p5 <- ggplot(ord) +
    geom_segment(aes(x = i, xend = i, y = baseline_r, yend = transfer_r),
                 colour = GRID, linewidth = 0.4) +
    geom_point(aes(i, baseline_r, colour = lab), size = 0.9) +
    geom_point(aes(i, transfer_r, colour = "Borzoi-TL"), size = 0.9) +
    scale_colour_manual(values = pal) +
    labs(title = sprintf("Every target, sorted by %s performance", lab),
         subtitle = "each vertical stem links one target's two scores",
         x = "target track (sorted)", y = "Pearson R") +
    theme_bench()
  save_both(p5, paste0(key, "_paired"), w = 8.0, h = 4.4)

  # 6. biological plausibility ---------------------------------------------
  mtf <- tf_of(df$match_track, key)
  df$plaus <- vapply(seq_len(nrow(df)),
                     function(i) classify(toupper(trimws(df$target[i])), mtf[i]), character(1))
  pl <- df |> count(plaus) |> mutate(plaus = reorder(plaus, n))
  p6 <- ggplot(pl, aes(n, plaus)) +
    geom_col(fill = col, alpha = 0.85, width = 0.62) +
    geom_text(aes(label = n), hjust = -0.35, colour = INK, fontface = "bold", size = 3.6) +
    scale_x_continuous(expand = expansion(mult = c(0, 0.14))) +
    labs(title = sprintf("Is the matched %s track the same factor?", lab),
         subtitle = "best-match is selected purely by correlation — TF identity is not constrained",
         x = "number of target tracks", y = NULL) +
    theme_bench()
  save_both(p6, paste0(key, "_plausibility"), w = 6.6, h = 3.6)

  summary_rows[[key]] <- data.frame(
    baseline = lab, n = nrow(df),
    transfer_mean = mean(df$transfer_r), baseline_mean = mean(df$baseline_r),
    mean_delta = mu, median_delta = median(df$delta),
    wins = sum(df$winner == "Borzoi-TL"), losses = sum(df$winner == lab),
    ties = sum(df$winner == "tie"),
    wilcoxon_p = suppressWarnings(wilcox.test(df$transfer_r, df$baseline_r, paired = TRUE)$p.value),
    same_TF_matches = sum(df$plaus == "same TF")
  )
  cat(sprintf("  %-18s mean Δ=%+.4f  %d/%d/%d W/L/T  -> 6 figures\n",
              lab, mu, sum(df$winner == "Borzoi-TL"), sum(df$winner == lab), sum(df$winner == "tie")))
}

summ <- bind_rows(summary_rows)
write.csv(summ, file.path(out_dir, sprintf("summary_by_baseline_%s.csv", split)), row.names = FALSE)
print(summ, digits = 4)
cat(sprintf("\nWrote %d figures (PNG+PDF) and 1 table to %s/\n", 6 * length(BASELINES), out_dir))

File: plot_benchmark_comparison.py
"""
Figures for the four-way benchmark (Borzoi-TL vs pretrained Borzoi vs Enformer
vs AlphaGenome), drawn from results_benchmark/benchmark_comparison_{split}.csv.

Three figures, each answering one question:
  1. model_comparison   — how do the four models' per-track R distributions compare?
  2. paired_delta       — is Borzoi-TL's margin over each baseline real?
  3. leakage_scatter    — where do Borzoi-TL's losses concentrate, and why?

Palette is the validated categorical order blue/orange/aqua/violet (passes the
lightness-band, chroma, protan+deutan CVD, normal-vision and contrast checks on
the all-pairs list). Aqua sits slightly under the 3.0 contrast floor against the
surface, so every series is also direct-labeled and legended — identity is never
carried by color alone.

Usage:
  python plot_benchmark_comparison.py --split test
"""
from __future__ import annotations

import argparse
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

# --- validated categorical palette (light mode) -----------------------------
C_TL, C_BZ, C_AG, C_EF = "#2a78d6", "#eb6834", "#1baf7a", "#4a3aa7"
SURFACE = "#fcfcfb"
INK, INK2, GRID = "#0b0b0b", "#52514e", "#d9d8d4"

MODELS = [
    ("borzoi_tl_r", "Borzoi-TL\n(ours)", C_TL),
    ("borzoi_pretrained_r", "Borzoi\npretrained", C_BZ),
    ("alphagenome_r", "AlphaGenome", C_AG),
    ("enformer_r", "Enformer", C_EF),
]
TIE_TOL = 0.01


def parse_args():
    p = argparse.ArgumentParser()
    p.add_argument("--split", default="test", choices=["val", "test"])
    p.add_argument("--results-dir", default="./results_benchmark")
    return p.parse_args()


def style(ax):
    """Recessive axes: no chartjunk, grid behind the marks."""
    ax.set_facecolor(SURFACE)
    for s in ("top", "right"):
        ax.spines[s].set_visible(False)
    for s in ("left", "bottom"):
        ax.spines[s].set_color(GRID)
        ax.spines[s].set_linewidth(1.0)
    ax.tick_params(colors=INK2, labelsize=9, length=3, width=1.0)
    ax.set_axisbelow(True)


def same_experiment(row) -> bool:
    """Does pretrained Borzoi own a track for this exact TF *and* cell line?

    String heuristic over the matched track's description — good enough to
    separate 'the model memorised this assay' from 'the model generalised',
    and every call is auditable in the comparison CSV.
    """
    tf = str(row["target_factor"]).upper()
    cl = str(row["target_cell_line"]).upper()
    trk = str(row["borzoi_pretrained_track"]).upper()
    return bool(tf) and tf in trk and cl != "NAN" and cl.replace("-", "") in trk.replace("-", "")


def fig_model_comparison(df, out, split):
    fig, ax = plt.subplots(figsize=(7.2, 4.6), facecolor=SURFACE)
    style(ax)
    rng = np.random.default_rng(0)

    for i, (col, label, color) in enumerate(MODELS):
        v = df[col].to_numpy(float)
        # jittered strip: 183 paired points, thin and translucent so density reads
        ax.scatter(i + rng.uniform(-0.16, 0.16, len(v)), v, s=9, color=color,
                   alpha=0.30, linewidths=0, zorder=2)
        bp = ax.boxplot([v], positions=[i], widths=0.42, showfliers=False,
                        patch_artist=True, zorder=3)
        bp["boxes"][0].set(facecolor="white", edgecolor=color, linewidth=1.6, alpha=0.92)
        for part in ("whiskers", "caps"):
            for a in bp[part]:
                a.set(color=color, linewidth=1.4)
        bp["medians"][0].set(color=color, linewidth=2.2)
        # direct label: the mean, so identity never depends on color alone
        ax.text(i, v.max() + 0.035, f"{v.mean():.3f}", ha="center", va="bottom",
                fontsize=10, color=color, fontweight="bold")

    ax.set_xticks(range(len(MODELS)))
    ax.set_xticklabels([m[1] for m in MODELS], color=INK, fontsize=9.5)
    ax.set_ylabel("Pearson R per target track", color=INK, fontsize=10)
    ax.set_ylim(0, 1.02)
    ax.yaxis.grid(True, color=GRID, linewidth=0.8)
    ax.set_title(f"Per-track accuracy on {len(df)} held-out CRC TF ChIP-seq tracks",
                 color=INK, fontsize=11.5, pad=26, loc="left")
    ax.text(0, 1.055, "bold value = mean · box = median and IQR · dots = individual tracks",
            transform=ax.transAxes, fontsize=8.5, color=INK2)
    fig.tight_layout()
    for ext in ("png", "pdf"):
        fig.savefig(out / f"model_comparison_{split}.{ext}", dpi=200, facecolor=SURFACE)
    plt.close(fig)


def fig_paired_delta(df, out, split):
    from borzoi_baseline_utils import bootstrap_ci
    rows = []
    for col, label, color in MODELS[1:]:
        d = (df["borzoi_tl_r"] - df[col]).to_numpy(float)
        lo, hi = bootstrap_ci(d, np.mean)
        rows.append((label.replace("\n", " "), d.mean(), lo, hi, color,
                     int((d > TIE_TOL).sum()), int((d < -TIE_TOL).sum()),
                     int((np.abs(d) <= TIE_TOL).sum())))

    fig, ax = plt.subplots(figsize=(7.2, 3.5), facecolor=SURFACE)
    style(ax)
    ax.axvline(0, color=INK2, linewidth=1.2, zorder=1)
    for i, (label, m, lo, hi, color, w, l, t) in enumerate(rows):
        y = len(rows) - 1 - i
        ax.plot([lo, hi], [y, y], color=color, linewidth=2.4, solid_capstyle="round", zorder=3)
        ax.scatter([m], [y], s=90, color=color, zorder=4, linewidths=0)
        ax.text(hi + 0.004, y, f"  +{m:.3f}", va="center", fontsize=10,
                color=color, fontweight="bold")
        # caption sits directly under its own row, in data coords, so it stays
        # attached no matter how many comparisons are plotted
        ax.text(lo, y - 0.26, f"{w} win · {l} lose · {t} tie",
                ha="left", va="top", fontsize=8.5, color=INK2)

    ax.set_yticks(range(len(rows)))
    ax.set_yticklabels([r[0] for r in reversed(rows)], color=INK, fontsize=10)
    ax.set_xlabel("mean ΔR  (Borzoi-TL minus baseline, 95% bootstrap CI)",
                  color=INK, fontsize=10)
    ax.set_ylim(-0.55, len(rows) - 0.45)
    ax.set_xlim(-0.006, 0.075)          # headroom for the direct labels
    ax.xaxis.grid(True, color=GRID, linewidth=0.8)
    ax.set_title("Borzoi-TL's margin over each baseline", color=INK,
                 fontsize=11.5, pad=20, loc="left")
    ax.text(0, 1.06, "every interval excludes zero — the ordering is not sampling noise",
            transform=ax.transAxes, fontsize=8.5, color=INK2)
    fig.tight_layout()
    for ext in ("png", "pdf"):
        fig.savefig(out / f"paired_delta_{split}.{ext}", dpi=200, facecolor=SURFACE)
    plt.close(fig)


def fig_leakage_scatter(df, out, split):
    leak = df.apply(same_experiment, axis=1).to_numpy()
    fig, ax = plt.subplots(figsize=(6.0, 5.6), facecolor=SURFACE)
    style(ax)

    lim = (0.10, 0.95)
    ax.plot(lim, lim, color=INK2, linewidth=1.2, linestyle=(0, (5, 4)), zorder=1)
    ax.scatter(df.loc[~leak, "borzoi_pretrained_r"], df.loc[~leak, "borzoi_tl_r"],
               s=34, color=C_TL, alpha=0.75, linewidths=0.5, edgecolors="white",
               zorder=3, label=f"no matching assay  (n={int((~leak).sum())})")
    ax.scatter(df.loc[leak, "borzoi_pretrained_r"], df.loc[leak, "borzoi_tl_r"],
               s=34, color=C_BZ, alpha=0.85, linewidths=0.5, edgecolors="white",
               zorder=4, label=f"same TF + cell line  (n={int(leak.sum())})")

    d_leak = (df.loc[leak, "borzoi_tl_r"] - df.loc[leak, "borzoi_pretrained_r"]).mean()
    d_gen = (df.loc[~leak, "borzoi_tl_r"] - df.loc[~leak, "borzoi_pretrained_r"]).mean()
    ax.text(0.035, 0.965, f"mean ΔR when Borzoi has the same assay:  {d_leak:+.3f}\n"
                          f"mean ΔR when it does not:                {d_gen:+.3f}",
            transform=ax.transAxes, fontsize=9, color=INK, va="top", family="monospace",
            bbox=dict(boxstyle="round,pad=0.5", facecolor="white", edgecolor=GRID))
    # sits in the empty band above the diagonal, clear of the stats box and legend
    ax.text(0.05, 0.62, "above the line =\nBorzoi-TL better", transform=ax.transAxes,
            ha="left", fontsize=8.5, color=INK2, style="italic")

    ax.set_xlim(lim); ax.set_ylim(lim)
    ax.set_xlabel("pretrained Borzoi best-match R", color=INK, fontsize=10)
    ax.set_ylabel("Borzoi-TL R", color=INK, fontsize=10)
    ax.grid(True, color=GRID, linewidth=0.8)
    ax.set_title("Where Borzoi-TL loses — and why", color=INK, fontsize=11.5,
                 pad=28, loc="left")
    ax.text(0, 1.035, "losses concentrate where pretrained Borzoi already saw the same experiment",
            transform=ax.transAxes, fontsize=8.5, color=INK2)
    leg = ax.legend(loc="lower right", frameon=True, fontsize=8.5)
    leg.get_frame().set(edgecolor=GRID, facecolor="white")
    fig.tight_layout()
    for ext in ("png", "pdf"):
        fig.savefig(out / f"leakage_scatter_{split}.{ext}", dpi=200, facecolor=SURFACE)
    plt.close(fig)


def main():
    args = parse_args()
    out = Path(args.results_dir)
    df = pd.read_csv(out / f"benchmark_comparison_{args.split}.csv")

    fig_model_comparison(df, out, args.split)
    fig_paired_delta(df, out, args.split)
    fig_leakage_scatter(df, out, args.split)

    # the stratification is a result, not just a picture — persist it as a table
    df["same_experiment_in_borzoi"] = df.apply(same_experiment, axis=1)
    df["delta_vs_borzoi_pretrained"] = df["borzoi_tl_r"] - df["borzoi_pretrained_r"]
    strat = df.groupby("same_experiment_in_borzoi").agg(
        n=("borzoi_tl_r", "size"),
        borzoi_tl_mean_r=("borzoi_tl_r", "mean"),
        borzoi_pretrained_mean_r=("borzoi_pretrained_r", "mean"),
        mean_delta=("delta_vs_borzoi_pretrained", "mean"),
        n_tl_loses=("delta_vs_borzoi_pretrained", lambda s: int((s < -TIE_TOL).sum())),
    ).round(4)
    strat.to_csv(out / f"stratified_by_assay_overlap_{args.split}.csv")

    df.sort_values("delta_vs_borzoi_pretrained").head(40).to_csv(
        out / f"borzoi_tl_losses_{args.split}.csv", index=False)

    print(strat.to_string())
    print(f"\nWrote 3 figures (PNG+PDF) and 2 tables to {out}/")


if __name__ == "__main__":
    main()
